lib/Conversion/DistToStandard/DistToStandard.cpp

```diff
@@ -102,79 +123,168 @@ struct RuntimePrototypesOpConverter
+          loc, sz,
+          builder.create<mlir::arith::SubIOp>(loc, np,
+                                              createInt(loc, builder, 1))),
+      np);
+}
+
+/// Convert ::imex::dist::LocalOffsetsOp into shape and arith calls.
+/// We currently assume evenly split data.
 struct LocalOffsetsOpConverter
-    : public mlir::OpRewritePattern<::imex::dist::LocalOffsetsOp> {
-  using OpRewritePattern::OpRewritePattern;
+    : public mlir::OpConversionPattern<::imex::dist::LocalOffsetsOp> {
+  using OpConversionPattern::OpConversionPattern;
 
   ::mlir::LogicalResult
   matchAndRewrite(::imex::dist::LocalOffsetsOp op,
-                  ::mlir::PatternRewriter &rewriter) const override {
-    // get guid and rank and call runtime function
+                  ::imex::dist::LocalOffsetsOp::Adaptor adaptor,
+                  ::mlir::ConversionPatternRewriter &rewriter) const override {
+    // FIXME: non-even partitions, ndims
     auto loc = op.getLoc();
-    rewriter.replaceOp(
-        op, createCallGetRankedData(loc, rewriter, "_idtr_local_offsets",
-                                    op.getPtensor(), op.getRank()));
+    auto converter = *getTypeConverter();
+
+    auto sz0_ = rewriter.create<::mlir::tensor::ExtractOp>(
+        loc, adaptor.getGshape(),
+        ::mlir::ValueRange({createIndex(loc, rewriter, 0)}));
+    auto i64Typ = rewriter.getI64Type();
+    auto sz0 =
+        sz0_.getResult().getType() == i64Typ
+            ? sz0_.getResult()
+            : rewriter.create<::mlir::arith::IndexCastOp>(loc, i64Typ, sz0_)
+                  .getResult();
+    auto tsz = createTileSize(loc, rewriter, sz0, adaptor.getNumProcs());
+    auto off =
+        rewriter.create<mlir::arith::MulIOp>(loc, adaptor.getPrank(), tsz);
+    rewriter.replaceOpWithNewOp<::mlir::tensor::FromElementsOp>(
+        op, converter.convertType(op.getType()), off.getResult());
+
     return ::mlir::success();
   }
 };
 
-// LocalShapeOp -> call into _idtr_local_shape
+/// Convert ::imex::dist::LocalShapeOp into shape and arith calls.
+/// We currently assume evenly split data.
 struct LocalShapeOpConverter
-    : public mlir::OpRewritePattern<::imex::dist::LocalShapeOp> {
-  using OpRewritePattern::OpRewritePattern;
+    : public mlir::OpConversionPattern<::imex::dist::LocalShapeOp> {
+  using OpConversionPattern::OpConversionPattern;
 
   ::mlir::LogicalResult
   matchAndRewrite(::imex::dist::LocalShapeOp op,
-                  ::mlir::PatternRewriter &rewriter) const override {
-    // get guid and rank and call runtime function
+                  ::imex::dist::LocalShapeOp::Adaptor adaptor,
+                  ::mlir::ConversionPatternRewriter &rewriter) const override {
+    // FIXME: non-even partitions, ndims
     auto loc = op.getLoc();
-    rewriter.replaceOp(
-        op, createCallGetRankedData(loc, rewriter, "_idtr_local_shape",
-                                    op.getPtensor(), op.getRank()));
+    auto converter = *getTypeConverter();
```

Comment:
This will do a full copy of `converter ` which is probably not what you wanted.

Reply:
Fixed.